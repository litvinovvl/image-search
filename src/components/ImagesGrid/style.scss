.container {
  padding: 40px 80px;
  background-color: #3A71D5;
  height: 100vh;
  overflow: auto;
  transition: height 1s ease-in-out;

  &.hidden {
    height: 0;
    padding: 0;
  }

  .images-wrapper {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: 300px;
    grid-gap: 30px;
  
    & .image {
      cursor: pointer;
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }

  .spinner-wrapper {
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    color: #FFF;
    font-size: 32px;
    font-weight: bold;
  }
}

.modal {
  position: absolute;
  top: 40px;
  left: 40px;
  right: 40px;
  bottom: 40px;
  background-color: #353335;

  .modal-content {
    width: 100%;
    height: 100%;

    .image {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }
}

.overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
}
