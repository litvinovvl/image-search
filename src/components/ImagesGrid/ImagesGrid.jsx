import React, { useState } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';

import { ReactComponent as NoResultsIcon } from '../../assets/images/sad-face.svg';
import { ReactComponent as LoadingIcon } from '../../assets/images/think-face.svg';

import './style.scss';

const ImagesGrid = ({
  images, isLoading, error,
}) => {
  const [openedImage, setOpenedImage] = useState(null);
  const isModalOpened = Boolean(openedImage);

  const handleImgClick = (image) => () => setOpenedImage(image);

  const handleImgKeyPress = (image) => (e) => {
    if (e.key === 'Enter') setOpenedImage(image);
  };

  const renderImages = () => {
    if (error) {
      return (
        <div className="spinner-wrapper">
          <NoResultsIcon width={48} hanging={48} />
          Something went wrong, try again
        </div>
      );
    }

    if (isLoading) {
      return (
        <div className="spinner-wrapper">
          <LoadingIcon width={48} hanging={48} />
          Loading
        </div>
      );
    }

    if (images && images.length) {
      return (
        <div className="images-wrapper">
          {
            images.map((image) => (
              <div
                key={image.id}
                role="button"
                onClick={handleImgClick(image)}
                onKeyPress={handleImgKeyPress(image)}
                tabIndex={0}
              >
                <img
                  src={image.webformatURL}
                  alt={image.tags}
                  className="image"
                />
              </div>

            ))
          }
        </div>
      );
    }

    return (
      <div className="spinner-wrapper">
        <NoResultsIcon width={48} hanging={48} />
        No results found!
      </div>
    );
  };

  const closeModal = () => setOpenedImage(null);

  return (
    <div className={classNames('container', { hidden: !images })}>
      {renderImages()}
      <ReactModal
        className="modal"
        overlayClassName="overlay"
        isOpen={isModalOpened}
        onRequestClose={closeModal}
      >
        {isModalOpened && (
          <div className="modal-content">
            <img src={openedImage.largeImageURL} alt={openedImage.tags} className="image" />
          </div>
        )}
      </ReactModal>
    </div>
  );
};

ImagesGrid.defaultProps = {
  images: null,
  isLoading: false,
  error: false,
};

ImagesGrid.propTypes = {
  images: PropTypes.arrayOf(PropTypes.object),
  isLoading: PropTypes.bool,
  error: PropTypes.bool,
};

export default ImagesGrid;
