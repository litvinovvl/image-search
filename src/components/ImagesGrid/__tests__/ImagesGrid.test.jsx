import React from 'react';
import ReactModal from 'react-modal';
import { shallow } from 'enzyme';

import ImagesGrid from '../ImagesGrid';

const image = { largeImageURL: 'large_url', webformatURL: 'test_url', tags: 'tags' };
const images = [
  { ...image, id: 1 },
  { ...image, id: 2 },
];

const props = {
  isLoadding: false,
  error: false,
  images,
};

describe('App', () => {
  beforeEach(jest.clearAllMocks);

  it('should render ImagesGrid properly initially', () => {
    const component = shallow(<ImagesGrid {...props} images={null} />);

    expect(component).toMatchSnapshot();
  });

  it('should render ImagesGrid properly when loading', () => {
    const component = shallow(<ImagesGrid {...props} isLoading />);

    expect(component).toMatchSnapshot();
  });

  it('should render ImagesGrid properly when error', () => {
    const component = shallow(<ImagesGrid {...props} error />);

    expect(component).toMatchSnapshot();
  });

  it('should render ImagesGrid properly when no images found', () => {
    const component = shallow(<ImagesGrid {...props} images={[]} />);

    expect(component).toMatchSnapshot();
  });

  it('should render ImagesGrid properly when images loaded', () => {
    const component = shallow(<ImagesGrid {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('should pass initial props to modal properly', () => {
    const component = shallow(<ImagesGrid {...props} />);
    const modal = component.find(ReactModal);

    expect(modal.prop('className')).toBe('modal');
    expect(modal.prop('overlayClassName')).toBe('overlay');
    expect(modal.prop('isOpen')).toBe(false);
    expect(modal.prop('onRequestClose')).toEqual(expect.any(Function));
    expect(modal.prop('children')).toBe(false);
  });

  it('should render big image in modal on image click', () => {
    const component = shallow(<ImagesGrid {...props} />);

    component.find('div[role="button"]').at(0).simulate('click');
    component.update();

    const modal = component.find(ReactModal);
    expect(modal.prop('isOpen')).toBe(true);
    expect(modal.prop('children')).toMatchSnapshot();
  });

  it('should render big image in modal on image press enter', () => {
    const component = shallow(<ImagesGrid {...props} />);

    component.find('div[role="button"]').at(0).simulate('keyPress', { key: 'Enter' });
    component.update();

    const modal = component.find(ReactModal);
    expect(modal.prop('isOpen')).toBe(true);
    expect(modal.prop('children')).toMatchSnapshot();
  });

  it('should not render big image in modal on image press not enter', () => {
    const component = shallow(<ImagesGrid {...props} />);

    component.find('div[role="button"]').at(0).simulate('keyPress', { key: 'Shift' });
    component.update();

    const modal = component.find(ReactModal);
    expect(modal.prop('isOpen')).toBe(false);
    expect(modal.prop('children')).toBe(false);
  });
});
