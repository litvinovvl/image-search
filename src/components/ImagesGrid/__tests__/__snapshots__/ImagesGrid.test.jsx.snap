// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render ImagesGrid properly initially 1`] = `
<div
  className="container hidden"
>
  <div
    className="spinner-wrapper"
  >
    <ForwardRef(SvgSadFace)
      hanging={48}
      width={48}
    />
    No results found!
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  />
</div>
`;

exports[`App should render ImagesGrid properly when error 1`] = `
<div
  className="container"
>
  <div
    className="spinner-wrapper"
  >
    <ForwardRef(SvgSadFace)
      hanging={48}
      width={48}
    />
    Something went wrong, try again
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  />
</div>
`;

exports[`App should render ImagesGrid properly when images loaded 1`] = `
<div
  className="container"
>
  <div
    className="images-wrapper"
  >
    <div
      key="1"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <img
        alt="tags"
        className="image"
        src="test_url"
      />
    </div>
    <div
      key="2"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <img
        alt="tags"
        className="image"
        src="test_url"
      />
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  />
</div>
`;

exports[`App should render ImagesGrid properly when loading 1`] = `
<div
  className="container"
>
  <div
    className="spinner-wrapper"
  >
    <ForwardRef(SvgThinkFace)
      hanging={48}
      width={48}
    />
    Loading
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  />
</div>
`;

exports[`App should render ImagesGrid properly when no images found 1`] = `
<div
  className="container"
>
  <div
    className="spinner-wrapper"
  >
    <ForwardRef(SvgSadFace)
      hanging={48}
      width={48}
    />
    No results found!
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={0}
    contentElement={[Function]}
    isOpen={false}
    onRequestClose={[Function]}
    overlayClassName="overlay"
    overlayElement={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    preventScroll={false}
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  />
</div>
`;

exports[`App should render big image in modal on image click 1`] = `
<div
  className="modal-content"
>
  <img
    alt="tags"
    className="image"
    src="large_url"
  />
</div>
`;

exports[`App should render big image in modal on image press enter 1`] = `
<div
  className="modal-content"
>
  <img
    alt="tags"
    className="image"
    src="large_url"
  />
</div>
`;
